docs/cookbook/rag.md

```diff
@@ -1,7 +1,424 @@
+
+from typing import Optional
+
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+MODEL = "gpt-3.5-turbo"
+EMBEDDINGS_MODEL = "text-embedding-ada-002"
+MAX_TOKENS = 1000
+TEXT_COLUMN = "Text"
+EMBEDDINGS_COLUMN = "embeddings"
+FILENAME = "news_article_dataset.pkl"
+URL = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+PINECONE_INDEX = "news-articles"
+PINECONE_NAMESPACE = "articles"
+
+class Settings(BaseSettings):
+    openai_api_key: Optional[str] = None
+    pinecone_api_key: Optional[str] = None
+
+    model_config = SettingsConfigDict(env_file=".env")
+```
+
+```python
+# .env
+
+OPENAI_API_KEY = "YOUR_OPENAI_API_KEY"
+PINECONE_API_KEY = "YOUR_PINECONE_API_KEY"
+```
+
+## Load and preprocess the data
+
+Before we load raw data into a pandas `Dataframe`, we have to handle token limits for both chat and embeddings models. In our case, where `gpt-3.5-turbo` has a lower token limit (4096) than the `text-embedding-ada-002` (8191), the chat becomes the limiting factor. A crude solution is to chunk each article into equal token lengths of at most `MAX_TOKENS=1000`, so we may fit 3 article snippets as well as any text in our hand-written section of the prompt. The function `split_text()` below contains the actual chunking logic.
+
+For counting the token lengths, we will use [tiktoken](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) since we will be using OpenAI for both the chat and embedding models. [tiktoken](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) is a useful library provided by OpenAI for encoding strings and decoding tokens for their models.
+
+```python
+# utils.py
+
+import pandas as pd
+import tiktoken
+from config import MODEL, TEXT_COLUMN,
+
+def load_data(url: str, max_tokens: int) -> pd.DataFrame:
+    """Loads data from a url after splitting larger texts into smaller chunks.
+    Args:
+        url: the url to load the data from.
+        max_tokens: the maximum number of tokens per chunk.
+    Returns:
+        The dataframe with the data from the url.
+    """
+    df = pd.read_csv(url)
+    split_articles = []
+    encoder = tiktoken.encoding_for_model(MODEL)
+    for i, row in df.iterrows():
+        text = row[TEXT_COLUMN]
+        tokens = encoder.encode(text)
+        if len(tokens) > max_tokens:
+            split_articles += split_text(text, tokens, max_tokens)
+            df.drop(i, inplace=True)
+
+    # Long texts which were dropped from the dataframe are now readded.
+    df = pd.concat(
+        [df, pd.DataFrame(split_articles, columns=[TEXT_COLUMN])], ignore_index=True
+    )
+
+    return df
+
+
+def split_text(text: str, tokens: list[int], max_tokens: int) -> list[str]:
+    """Roughly splits a text into chunks according to max_tokens.
+    Text is split into equal word counts, with number of splits determined by how many
```

Comment:
nit: add missing newline here between first sentence and additional information.